walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -11,13 +11,130 @@
+    eventAttendees = event.getAttendees();
+    boolean isOverlappedAttendee = false;
+    for(String attendee:eventAttendees){
+      if(attendees.contains(attendee)){
+        isOverlappedAttendee = true;
+        // add a break to terminate loop early if at least one attendee is present at event
+        break; 
+        }
+      }
+      if(isOverlappedAttendee){
+        TimeRange blocked = event.getWhen();
+        for(TimeRange timeslot: fullDay){
+          if(blocked.overlaps(timeslot)){
+            // case one: event is fully enclosed within a free timeslot
+            if(blocked.start() > timeslot.start() && blocked.end() < timeslot.end()){
+              freeTimes.add(TimeRange.fromStartEnd(timeslot.start(), blocked.start(),false));
+              freeTimes.add(TimeRange.fromStartEnd(blocked.end(),timeslot.end(),false));
+            }
+            // case two: event overlaps end of free timeslot
+            else if(blocked.start() >= timeslot.start() && blocked.end() >= timeslot.end()){
+              if(blocked.start()  != timeslot.start()){
+                freeTimes.add(TimeRange.fromStartEnd(timeslot.start(), blocked.start(),false));
+              }
+            }
+            // case three: event overlaps start of free timeslot
+            else if(blocked.start() <= timeslot.start() && blocked.end() >= timeslot.start()){
+              if(blocked.end()  != timeslot.end()){
+                freeTimes.add(TimeRange.fromStartEnd(blocked.end(),timeslot.end(),false));
+              }
+            }
+            blockedTimes.add(timeslot);
+          }
+        }
+        fullDay.addAll(freeTimes);
+        fullDay.removeAll(blockedTimes);
+        // must clear everytime so that blocked events don't get added back
+        freeTimes.clear();
+        blockedTimes.clear();
+      }
+    }
+        
+    // filter through available times to find sufficient durations
+    for(TimeRange available: fullDay){
+      if(available.duration() < request.getDuration()){
+          blockedTimes.add(available);
+      }
+    }
+    fullDay.removeAll(blockedTimes);
+    
+    // try to insert optional attendees into existing timeslots 
+    attendees = request.getOptionalAttendees();
+    for(Event event: events){
+        eventAttendees = event.getAttendees();
+        boolean isOverlappedOptAttendee = false;
+        for(String attendee:eventAttendees){
+            if(attendees.contains(attendee)){
+              isOverlappedOptAttendee = true;
+              break; 
+            }
+        }
+        if(isOverlappedOptAttendee){
+          TimeRange blocked = event.getWhen();
+              int start = blocked.start();
+              int end = blocked.end();
+              for(TimeRange timeslot: fullDay){
+                if(blocked.overlaps(timeslot) && timeslot.duration()-blocked.duration() >= request.getDuration()){
+                    if(blocked.start() > timeslot.start() && blocked.end() < timeslot.end()){
```

Comment:
It would be good to break up this method into several helper methods, especially these nested if statements, so that the code would be more readable. Just copying this section into a method called `addOverlappedOptionalAttnedee` or something would immediately improve readability.

Reply:
Done.